design-proposals/sdn-ipam-secondary-nets.md

```diff
@@ -0,0 +1,450 @@
+The IP addresses allocated to the VM will **not** be released to the IP pool
+when there's a corresponding `IPAMClaim` for the attachment being deleted.
+
+#### Starting a (previously stopped) Virtual Machine
+This flow is - from a CNI perspective - quite similar to the
+[Creating a VM flow](#creating-a-virtual-machine):
+1. the workload controller (KubeVirt) templates the pod, featuring the required
+   `IPAMClaim` references in each network selection element.
+2. OVN-Kubernetes will read the network selection element; if it sees an
+   `IPAMClaim` reference, it will attempt to read it from the datastore.
+3. if the read `IPAMClaim` features IP addresses in its status, OVN-Kubernetes
+   will configure the interface with those IPs. If the status doesn't feature
+   IP addresses, the CNI plugin will retry later, with exponential backoff.
+
+```mermaid
+sequenceDiagram
+
+  actor user
+  participant KubeVirt
+  participant Kubernetes
+  participant CNI
+
+  user->>KubeVirt: startVM(vmName)
+
+  note over KubeVirt: podName := "launcher-<vm name>"
+  KubeVirt->>Kubernetes: createPod(name=podName, ipam-claims=[claimName])
+
+  Kubernetes->>CNI: CNI ADD
+  note over CNI: this CNIs has an IPAMClaims informer
+
+  CNI->>CNI: ipamClaim := readIPAMClaim(claimName)
+  CNI->>CNI: configIface(ipamClaim.Status.IPs)
+
+  CNI-->>Kubernetes: OK
+
+  Kubernetes-->>KubeVirt: OK
+
+
+  KubeVirt-->>user: OK
+```
+
+When creating the pod, the OVN-Kubernetes IPAM module finds existing
+`IPAMClaim`s for the workload. It will thus use those already reserved
+allocations, instead of generating brand new allocations for the pod where the
+encapsulating object will run. The migration scenario is similar.
+
+#### Removing a Virtual Machine
+This flow is - from a CNI perspective - quite similar to the
+[Stopping a VM flow](#stopping-a-virtual-machine). The main difference is after
+the VM is deleted, Kubernetes Garbage Collection will kick in, and remove the
+orphaned `IPAMClaim`s. OVN-Kubernetes will need to react to the `IPAMClaim`
+delete event, to return those IP addresses to the respective IP pools.
+
+```mermaid
+sequenceDiagram
+
+  actor user
+  participant KubeVirt
+  participant Kubernetes
+  participant CNI
+  participant IPAM CNI
+
+  user->>KubeVirt: deleteVM(vmName)
+
+  KubeVirt->>Kubernetes: deletePod(launcher-vmName)
+
+  Kubernetes->>CNI: CNI DEL
+  CNI-->>Kubernetes: OK
+
+  Kubernetes-->>KubeVirt: OK
+
+  KubeVirt-->>user: OK
+
+  loop for every orphaned IPAMClaim
+    Kubernetes->>Kubernetes: deleteIPAMClaim()
+  end
+
+  note over IPAM CNI: this CNI has an IPAMClaim informer
+
+Kubernetes->>IPAM CNI: removeEvent(ipamClaim)
+IPAM CNI->>IPAM CNI: returnToPool(ipamClaim.Status.IPs)
+```
+
+#### Hot-plug a VM interface
+This flow is exactly the same as
+[starting a VM](#starting-a-previously-stopped-virtual-machine).
+
+#### Hot-unplug a VM interface
+Hot-unplugging an interface from a KubeVirt VM is a two step process, requiring
```

Comment:
I guess if you do that you'll pretty much leak the pod interface right ? I'm assuming we actually clean up the resources properly.

Reply:
What does it mean  "leak the pod interface"?
The unplug request arrives from the VM spec, passed to the VMI, and when everything is removed (from the pod netns & domain) we delete the interface entry.

Reply:
The pod network annotation is updated and the relevant interface/network entry is removed, which in turn the CNI is called to remove it. In parallel Kubevirt removes the configuration in the pod netns.
When the network interface is no longer appearing the pod annotation (I think in the netowrk-status annotation), in the domain and in the pod netns, we consider the interface gone completely.

From an IPAM perspective, I think the removal of the network-status annotation entry is enough to consider it released.

Reply:
> The pod network annotation is updated and the relevant interface/network entry is removed, which in turn the CNI is called to remove it. In parallel Kubevirt removes the configuration in the pod netns. When the network interface is no longer appearing the pod annotation (I think in the netowrk-status annotation), in the domain and in the pod netns, we consider the interface gone completely.

Ok, but IIUC, for CNI to be invoked, you're relying on a dynamic component to invoke the CNI delete (like the [multus dynamic networks controller](https://github.com/k8snetworkplumbingwg/multus-dynamic-networks-controller), right ? 

Keep in mind we do not ship that downstream, so, essentially, on downstream, unless there's a migration, we're effectively leaking that interface - since we're only unplugging from the domain.

Right ? 

I guess this was copied verbatim from the downstream enhancement proposal, and does not apply well to the upstream design doc.

> 
> From an IPAM perspective, I think the removal of the network-status annotation entry is enough to consider it released.

Reply:
> Ok, but IIUC, for CNI to be invoked, you're relying on a dynamic component to invoke the CNI delete (like the [multus dynamic networks controller](https://github.com/k8snetworkplumbingwg/multus-dynamic-networks-controller), right ?

Yes.
It is also the case with hotplug.

> Keep in mind we do not ship that downstream, so, essentially, on downstream, unless there's a migration, we're effectively leaking that interface - since we're only unplugging from the domain.
> 
> Right ?

In this sense you are right.
Usually D/S solutions prefer a complete e2e solution and without this it will never be complete.
I mean, the need to manually migrate a VM is not that nice. I would expect D/S projects/products to use the full on-the-fly solution.